### PR TITLE
re-org safety: preparatory commits

### PR DESCRIPTION
**Pull request by guggero**
Depends on https://github.com/lightningnetwork/lnd/pull/7859.

Spun out of https://github.com/lightninglabs/taproot-assets/pull/419 in order to reduce the final diff.
All commits are mostly related to re-org safety but might appear out of context without also taking a look at #419.

### CODE REVIEW

File: proof/archive.go [hunk 1]
```diff
@@ -214,6 +219,12 @@ func (f *FileArchiver) ImportProofs(_ context.Context,
 			return err
 		}
 
+		// Can't replace a file that doesn't exist yet.
+		if replace && !lnrpc.FileExists(proofPath) {
```
Comment (Roasbeef) [inline proof/archive.go:223]: 👍

File: tapdb/assets_store.go [hunk 2]
```diff
@@ -1329,22 +1329,67 @@ func (a *AssetStore) importAssetFromProof(ctx context.Context,
...
 	proofs ...*proof.AnnotatedProof) error {
 
 	var writeTxOpts AssetStoreTxOptions
 	err := a.db.ExecTx(ctx, &writeTxOpts, func(q ActiveAssetsStore) error {
 		for _, p := range proofs {
-			err := a.importAssetFromProof(ctx, q, p)
-			if err != nil {
-				return fmt.Errorf("unable to import asset: %w",
-					err)
+			if replace {
```
Comment (Roasbeef) [inline tapdb/assets_store.go:1381]: Would be good to have a basic test for this in the db.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: proof/mint.go [hunk 3]
```diff
@@ -149,25 +153,20 @@ func NewMintingBlobs(params *MintParams,
 	}
 
 	ctx := context.Background()
-	blobs := make(AssetBlobs, len(proofs))
+
+	// Before we return the proofs, we'll verify that we actually generated
+	// valid ones.
```
Comment (ffranr) [inline proof/mint.go:158]: how about: `// Verify generated proofs.`

File: rpcserver.go [hunk 4]
```diff
@@ -1099,11 +1099,13 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
...
-	_, err = proofFile.Verify(
-		ctx, headerVerifier,
-	)
+	_, err = proofFile.Verify(ctx, headerVerifier)
 	valid := err == nil
 
+	if !valid {
+		rpcsLog.Errorf("Proof verification failed with err: %v", err)
+	}
+
```
Comment (ffranr) [inline rpcserver.go:1108]: This pattern is strange. I think we should get rid of the variable `valid` and this `if` clause can just inspect `err`.

When populating the `Valid` field in the `return` statement, we can just set it to `err == nil`.

File: server.go [hunk 5]
```diff
@@ -297,15 +297,18 @@ func (s *Server) RunUntilShutdown(mainErrChan <-chan error) error {
...
 	case err := <-mainErrChan:
 		if err == nil {
+			srvrLog.Debug("Main err chan closed")
 			return nil
 		}
 
 		// We'll report the error to the main daemon, but only if this
 		// isn't a context cancel.
 		if fn.IsCanceled(err) {
+			srvrLog.Debugf("Got context canceled error: %v", err)
```
Comment (ffranr) [inline server.go:311]: :+1:

**Review by ffranr — APPROVED** [review]
Nice changes!

**Review by jharveyb — APPROVED** [review]
Looks good! No further comment besides adding test coverage for the new `replace` option.

File: tapdb/assets_store.go (hunk 2 above)
Comment (guggero) [reply tapdb/assets_store.go:1381]: Makes sense, added.

**Comment by guggero** [comment]
PR is ready now, addressed all the review comments. Though we should wait for https://github.com/lightningnetwork/lnd/pull/7859 to be merged (updated that as well with a reproducing unit test).

**Comment by Roasbeef** [comment]
`lnd` PR has been merged!

**Comment by guggero** [comment]
Done, ready for final review.

**Review by Roasbeef — APPROVED** [review]
LGTM ☔️

**Pull request merged**